httpie/manager/plugins.py

```diff
@@ -0,0 +1,188 @@
+import sys
+import textwrap
+from collections import defaultdict
+from contextlib import suppress
+from pathlib import Path
+from typing import Optional, List
+
+import importlib_metadata
+
+from httpie.manager.cli import parser
+from httpie.compat import get_dist_name
+from httpie.context import Environment
+from httpie.status import ExitStatus
+
+
+class PluginInstaller:
+
+    def __init__(self, env: Environment, debug: bool = False) -> None:
+        self.env = env
+        self.dir = env.config.plugins_dir
+        self.debug = debug
+
+        self.setup_plugins_dir()
+
+    def setup_plugins_dir(self) -> None:
+        try:
+            self.dir.mkdir(
+                exist_ok=True,
+                parents=True
+            )
+        except OSError:
+            self.env.stderr.write(
+                f'Couldn\'t create "{self.dir!s}"'
+                ' directory for plugin installation.'
+                ' Please re-check the permissions for that directory,'
+                ' and if needed, allow write-access.'
+            )
+            raise
+
+    def fail(
+        self,
+        command: str,
+        target: Optional[str] = None,
+        reason: Optional[str] = None
+    ) -> ExitStatus:
+        message = f'Can\'t {command}'
+        if target:
+            message += f' {target!r}'
+        if reason:
+            message += f': {reason}'
+
+        self.env.stderr.write(message + '\n')
+        return ExitStatus.ERROR
+
+    def pip(self, *args, **kwargs) -> subprocess.CompletedProcess:
+        options = {
+            'check': True,
+            'shell': False,
+            'stdout': self.env.stdout,
+            'stderr': subprocess.PIPE,
+        }
+        options.update(kwargs)
+
+        cmd = [sys.executable, '-m', 'pip', *args]
+        return subprocess.run(
+            cmd,
+            **options
+        )
+
+    def install(self, targets: List[str]) -> Optional[ExitStatus]:
+        self.env.stdout.write(f"Installing {', '.join(targets)}...\n")
+        self.env.stdout.flush()
+
+        try:
+            self.pip(
+                'install',
+                f'--prefix={self.dir}',
+                '--no-warn-script-location',
+                *targets,
+            )
+        except subprocess.CalledProcessError as error:
+            reason = None
+            if error.stderr:
+                stderr = error.stderr.decode()
+
+                if self.debug:
+                    self.env.stderr.write('Command failed: ')
+                    self.env.stderr.write(' '.join(error.cmd) + '\n')
+                    self.env.stderr.write(textwrap.indent('  ', stderr))
+
+                last_line = stderr.strip().splitlines()[-1]
+                severity, _, message = last_line.partition(': ')
+                if severity == 'ERROR':
+                    reason = message
+
+            return self.fail('install', ', '.join(targets), reason)
+
+    def _uninstall(self, target: str) -> Optional[ExitStatus]:
```

Comment:
Only the main dep file gets uninstalled here, all the sub deps which it depends on are still present in the `site` directory. 

Eg: `httpie plugins uninstall httpie-nsof` would work perfectly as it has no dependencies (excluding `httpie`), whereas `httpie plugins uninstall httpie-oauth` wouldn't remove its depending deps such as `oauthlib`.

Reply:
This is simply how pip works, it is very problematic to backwards package dependency collection without storing extra metadata. It would bring a tremeounds amount of complexity, considering we would need to create a full-graph of dependency-relations and then remove them as well. As I mentioned, this is not a fully-fledged package manager but rather a simple overlay on top of pip.

Reply:
Hmm, it's worth adding a comment over there as a `TODO` probably... I think there should be a simple way of doing this, will give it a shot when I get some time.